extensions/wrapper/clients/typescript-client/src/oauth2/AccessTokenService.ts

```diff
@@ -0,0 +1,54 @@
+import {ClientCredentialsResponse, fetchClientCredentials} from './FetchUtils';
+import {buildTokenRequestFormData} from './RequestUtils';
+import {ClientCredentials} from './model/ClientCredentials';
+
+export class AccessTokenService {
+    private activeRequest: Promise<ClientCredentialsResponse> | null = null;
+    private refreshTimeout?: NodeJS.Timeout;
+    private tokenRequestFormData: string;
+    private accessToken: string | null = null;
+
+    constructor(private clientCredentials: ClientCredentials) {
+        this.tokenRequestFormData = buildTokenRequestFormData(
+            clientCredentials.clientId,
+            clientCredentials.clientSecret,
+        );
+    }
+
+    async getAccessToken(): Promise<string> {
+        if (!this.accessToken) {
+            return await this.refreshAccessToken();
+        }
+        return this.accessToken;
+    }
+
+    /**
+     * Synchronized refreshing of the access token
+     */
+    async refreshAccessToken(): Promise<string> {
+        if (this.activeRequest) {
+            await this.activeRequest;
+            return this.accessToken!;
+        }
+
+        this.accessToken = null;
+        this.activeRequest = fetchClientCredentials(
+            this.clientCredentials.tokenUrl,
+            this.tokenRequestFormData,
+        );
+        const clientCredentialsResponse = await this.activeRequest;
```

Comment:
Comment: Naming this here and in the function below just "response",  do you think that helps or not?